tests/integration/WC_Facebookcommerce_Integration_Test.php

```diff
@@ -771,8 +773,396 @@ public function test_is_product_sync_enabled_filter() {
+	}
+
+
+	/**
+	 * @see \WC_Facebookcommerce_Integration::fb_change_product_published_status()
+	 *
+	 * @param string $new_status
+	 * @param string $old_status
+	 * @param \WP_post $post
+	 *
+	 * @dataProvider provider_fb_change_product_published_status
+	 */
+	public function test_fb_change_product_published_status( $new_status, $old_status ) {
+
+		$product = $this->tester->get_product( [ 'status' => $new_status ] );
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'update_fb_visibility' => Expected::once(),
+		] );
+
+		$integration->fb_change_product_published_status( $new_status, $old_status, get_post( $product->get_id() ) );
+	}
+
+
+	/** @see test_fb_change_product_published_status() */
+	public function provider_fb_change_product_published_status() {
+
+		return [
+			'publish'   => [ 'publish', 'trash' ],
+			'unpublish' => [ 'trash', 'publish' ],
+		];
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::fb_change_product_published_status() */
+	public function test_fb_change_product_published_status_with_invalid_product() {
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'update_fb_visibility' => Expected::never(),
+		] );
+
+		$integration->fb_change_product_published_status( null, null, (object) [ 'ID' => 0 ] );
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_publish() */
+	public function test_on_product_publish_with_simple_product() {
+
+		$this->check_on_product_publish( $this->tester->get_product( [ 'status' => 'publish' ] ) );
+	}
+
+
+	/**
+	 * Tests that on_product_publish() delegates to the appropriate handler.
+	 *
+	 * @see \WC_Facebookcommerce_Integration::on_product_publish()
+	 *
+	 * @param \WC_Product $product product object
+	 */
+	private function check_on_product_publish( \WC_Product $product ) {
+
+		// on_product_publish() expects the Facebook Page ID to be configured
+		update_option( \WC_Facebookcommerce_Integration::SETTING_FACEBOOK_PAGE_ID, '123456' );
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'on_simple_product_publish'   => $product->is_type( 'variable' ) ? Expected::never() : Expected::once(),
+			'on_variable_product_publish' => $product->is_type( 'variable' ) ? Expected::once()  : Expected::never(),
+		] );
+
+		$integration->on_product_publish( $product->get_id() );
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_publish() */
+	public function test_on_product_publish_with_variable_product() {
+
+		$this->check_on_product_publish( $this->tester->get_variable_product( [
+			'children' => 3,
+			'status'   => 'publish',
+		] ) );
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_publish() */
+	public function test_on_product_publish_with_invalid_product_id() {
+
+		$this->check_on_product_publish_returns_early( 0 );
+	}
+
+
+	private function check_on_product_publish_returns_early( $product_id, \WC_Product $product = null ) {
```

Comment:
Same feedback as the other one, it would be a lot clearer if this used a provider.